sim/midas/src/main/scala/midas/widgets/ClockBridge.scala

```diff
@@ -147,13 +147,33 @@ class ClockTokenVector(numClocks: Int) extends Bundle with HasChannels with Cloc
 class ClockBridgeModule(params: ClockParameters)(implicit p: Parameters)
     extends BridgeModule[ClockTokenVector] {
   val clockInfo = params.clocks
+  val phaseRelationships = clockInfo map { cInfo => (cInfo.multiplier, cInfo.divisor) }
+
   lazy val module = new BridgeModuleImp(this) {
+
   val io = IO(new WidgetIO())
   val hPort = IO(new ClockTokenVector(clockInfo.size))
-  val phaseRelationships = clockInfo map { cInfo => (cInfo.multiplier, cInfo.divisor) }
+  val clocks = hPort.clocks
+
+  // Register which reports the number of remaining tokens to the driver.
+  val credit = RegInit(0.U(64.W))
+  attach(credit =/= 0.U, "HAS_CREDIT", ReadOnly)
+
+  // Register used to set the credit. Increments to the credit count are always accepted.
+  val creditIn = Wire(Decoupled(UInt(32.W)))
```

Comment:
I know this is totally inconsistent, but there's a member in Widget that is suitable for grabbing the width of the MMIO bus: https://github.com/firesim/firesim/blob/main/sim/midas/src/main/scala/midas/widgets/Widget.scala#L58